### PR TITLE
Fix DeepSpeed ZeRO-{1,2} for DPOTrainer

### PR DESCRIPTION
**Pull request by lewtun**
This PR fixes an issue that occurred when trying to run the DPO example with ZeRO-{1,2} via:

```shell
ACCELERATE_LOG_LEVEL=info TRANSFORMERS_VERBOSITY=info accelerate launch --config_file=examples/accelerate_configs/deepspeed_zero{1,2}.yaml examples/dpo.py
```

Which gave:

```
AssertionError: zero stage 2 requires an optimizer
Traceback (most recent call last):
  File "/fsx/lewis/git/trl/examples/dpo.py", line 154, in <module>
    dpo_trainer = DPOTrainer(
  File "/fsx/lewis/git/trl/trl/trainer/dpo_trainer.py", line 248, in __init__
    self.ref_model = self.accelerator._prepare_deepspeed(self.ref_model)[0]
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/accelerate/accelerator.py", line 1600, in _prepare_deepspeed
    engine, optimizer, _, lr_scheduler = deepspeed.initialize(**kwargs)
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/deepspeed/__init__.py", line 165, in initialize
    engine = DeepSpeedEngine(args=args,
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/deepspeed/runtime/engine.py", line 314, in __init__
    self.optimizer = self._configure_zero_optimizer(optimizer=None)
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/deepspeed/runtime/engine.py", line 1404, in _configure_zero_optimizer
    assert not isinstance(optimizer, DummyOptim), "zero stage {} requires an optimizer".format(zero_stage)
```

The problem can be traced to the use of `accelerator._prepare_deepspeed()` for the reference model, which requires an optimizer to be present for sharding. However, this isn't required for the reference model, so we can re-use the inference code I wrote for the `PPOTrainer` :)

Tested with:

```shell
ACCELERATE_LOG_LEVEL=info TRANSFORMERS_VERBOSITY=info accelerate launch --config_file=examples/accelerate_configs/deepspeed_zero{1,2,3}.yaml examples/scripts/dpo.py
```

**WandB for gpt2** 

* Accuracies match perfectly for ZeRO-{1,2} with a discrepancy in ZeRO-3 that was also observed in PPO (likely due to sharding)

Screenshot 2023-09-29 at 13 56 15

**Wandb for llama-2-7b**

* 1 node of 8 x A100 (80GB) OK for ZeRO-3
* 1 node of 8 x A100 (80GB) OOM for ZeRO-2
* 4 nodes of 8 x A100 (80GB) OK for ZeRO-2

### CODE REVIEW

**Comment by lewtun** [comment]
Hmm, I'm getting an issue with checkpoint saving now:

```
Traceback (most recent call last):
  File "/fsx/lewis/git/trl/examples/scripts/dpo.py", line 170, in <module>
    dpo_trainer.train()
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/transformers/trainer.py", line 1591, in train
    return inner_training_loop(
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/transformers/trainer.py", line 1984, in _inner_training_loop
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/transformers/trainer.py", line 2339, in _maybe_log_save_evaluate
    self._save_checkpoint(model, trial, metrics=metrics)
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/transformers/trainer.py", line 2471, in _save_checkpoint
    self.state.save_to_json(os.path.join(output_dir, TRAINER_STATE_NAME))
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/site-packages/transformers/trainer_callback.py", line 106, in save_to_json
    json_string = json.dumps(dataclasses.asdict(self), indent=2, sort_keys=True) + "\n"
  File "/fsx/lewis/miniconda/envs/trl/lib/python3.10/dataclasses.py", line 1238, in asdict
    return _asdict_inner(obj, dict_factory)
...
RecursionError: maximum recursion depth exceeded
```

Taking a look, but the code should still be fine to review

File: examples/accelerate_configs/deepspeed_zero1.yaml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 compute_environment: LOCAL_MACHINE
 debug: false
 deepspeed_config:
+  deepspeed_multinode_launcher: standard
```
Comment (lewtun) [inline examples/accelerate_configs/deepspeed_zero1.yaml:4]: This is needed to enable multi-node training

File: examples/scripts/dpo.py [hunk 2]
```diff
@@ -148,6 +148,7 @@ def split_prompt_and_responses(sample) -> Dict[str, str]:
         optim="rmsprop",
         warmup_steps=150,
         report_to=script_args.report_to,
+        bf16=True,
```
Comment (lewtun) [inline examples/scripts/dpo.py:151]: This is needed for compat with the deepspeed configs - I think it's also fine to have as a default in any case

File: examples/scripts/dpo.py
Comment (lewtun) [inline examples/scripts/dpo.py:1]: Note that I moved this script into it's proper place (not sure why it was in the root of `examples`)

Comment (kashif) [reply examples/scripts/dpo.py:1]: my bad!

**Review by lvwerra — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**